tracing-subscriber/src/filter/directive.rs

```diff
@@ -0,0 +1,364 @@
+    type Err = ParseError;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut split = s.split('=');
+        let part0 = split
+            .next()
+            .ok_or_else(|| ParseError::msg("string must not be empty"))?;
+
+        if let Some(part1) = split.next() {
+            let mut split = part0.split("[{");
+            let target = split.next().map(String::from);
+            let mut field_names = FilterVec::new();
+            if let Some(maybe_fields) = split.next() {
+                let fields = maybe_fields
+                    .strip_suffix("}]")
+                    .ok_or_else(|| ParseError::msg("expected fields list to end with '}]'"))?;
+                field_names.extend(fields.split(',').filter_map(|s| {
+                    if s.is_empty() {
+                        None
+                    } else {
+                        Some(String::from(s))
+                    }
+                }));
+            };
+            let level = part1.parse()?;
+            return Ok(Self {
+                level,
+                field_names,
+                target,
+            });
+        }
+
+        // Okay, the part after the `=` was empty, the directive is either a
+        // bare level or a bare target.
+        Ok(match part0.parse::<LevelFilter>() {
+            Ok(level) => Self {
+                level,
+                ..Default::default()
+            },
+            Err(_) => Self {
+                target: Some(String::from(part0)),
+                level: LevelFilter::TRACE,
+                ..Default::default()
+            },
+        })
+    }
+}
+
+// === impl ParseError ===
+
+impl ParseError {
+    pub(crate) fn new() -> Self {
+        ParseError {
+            kind: ParseErrorKind::Other(None),
+        }
+    }
+
+    pub(crate) fn msg(s: &'static str) -> Self {
+        ParseError {
+            kind: ParseErrorKind::Other(Some(s)),
+        }
+    }
+}
+
+impl fmt::Display for ParseError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self.kind {
+            ParseErrorKind::Other(None) => f.pad("invalid filter directive"),
+            ParseErrorKind::Other(Some(msg)) => write!(f, "invalid filter directive: {}", msg),
+            ParseErrorKind::Level(ref l) => l.fmt(f),
+            ParseErrorKind::Field(ref e) => write!(f, "invalid field filter: {}", e),
+        }
+    }
+}
+
+impl Error for ParseError {
+    fn description(&self) -> &str {
+        "invalid filter directive"
+    }
+
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        match self.kind {
+            ParseErrorKind::Other(_) => None,
+            ParseErrorKind::Level(ref l) => Some(l),
+            ParseErrorKind::Field(ref n) => Some(n.as_ref()),
+        }
+    }
+}
+
+impl From<Box<dyn Error + Send + Sync>> for ParseError {
+    fn from(e: Box<dyn Error + Send + Sync>) -> Self {
+        Self {
+            kind: ParseErrorKind::Field(e),
```

Comment:
This impl feels odd to me: it seems like it is only applicable to one, specific call-site, but it is very very general, as return type is a poor man’s anyhow basically. 

Ohhhhh, I now see that I can formulate a code-smell here: 


> having From<FooError> for BarError, which is used only once.

the single call-site means that we can convert from Foo to Bar *in one specific context*. But the impl means that we can do that in any context. I think, in such situations it’s better to do the conversion manually.


This is philosophical, but what is somewhat more well-grounded is that `ParseError` is a public type, so this would be a public impl which the user could use. And such an impl is a very odd one for a public API.

Reply:
again, this is code that previously existed, but was moved and renamed. I believe that this impl actually exists _because_ a user requested the ability to create custom `ParseError`s for some kind of testing purpose.